File: src/main/java/com/avs/filmoteca/data/ws/ApiClient.java
package com.avs.filmoteca.data.ws;

import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.converter.scalars.ScalarsConverterFactory;

public class ApiClient {

	public static FilmotecaInterface getFilmotecaInterface() {
		return getRetrofit(Environment.FILMOTECA_URL).create(FilmotecaInterface.class);
	}

	public static BackendInterface getBackendInterface() {
		return getRetrofit(Environment.BACKEND_URL).create(BackendInterface.class);
	}

	public static FCMInterface getFCMInterface() {
		return getRetrofit(Environment.FIREBASE_URL).create(FCMInterface.class);
	}

	private static Retrofit getRetrofit(String url) {
		return new Retrofit.Builder().baseUrl(url).addCallAdapterFactory(RxJavaCallAdapterFactory.create())
				.addConverterFactory(ScalarsConverterFactory.create())
				.addConverterFactory(GsonConverterFactory.create()).build();
	}

}

File: src/main/java/com/avs/filmoteca/data/ws/ApiClient.kt
package com.avs.filmoteca.data.ws

import com.avs.filmoteca.data.ws.security.BasicAuthCredentials
import com.avs.filmoteca.data.ws.security.BasicAuthInterceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.converter.scalars.ScalarsConverterFactory

object ApiClient {

    val filmotecaInterface: FilmotecaInterface by lazy {
        getRetrofit(Endpoint.Web).create(FilmotecaInterface::class.java)
    }

    val backendInterface: BackendInterface by lazy {
        getRetrofit(Endpoint.Service).create(BackendInterface::class.java)
    }

    val fcmInterface: FCMInterface by lazy {
        getRetrofit(Endpoint.Firebase).create(FCMInterface::class.java)
    }

    private val basicAuthHttpClient: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .addInterceptor(BasicAuthInterceptor(BasicAuthCredentials.user, BasicAuthCredentials.password))
            .build()
    }

    private fun getRetrofit(endpoint: Endpoint): Retrofit =
        Retrofit.Builder().apply {
            baseUrl(endpoint.endpoint)

            if (endpoint == Endpoint.Service)
                client(basicAuthHttpClient)

            addCallAdapterFactory(RxJavaCallAdapterFactory.create())
            addConverterFactory(ScalarsConverterFactory.create())
            addConverterFactory(GsonConverterFactory.create())
        }.build()

}
